### PR TITLE
Fix race condition in GetAppdomainStaticAddress test

### PR DESCRIPTION
**Pull request by davmason**
I have seen two intermittent failure modes for this test.

1. The thread that sits and does GCs can AV because there is a race with shutdown. When shutdown sets pCorProfilerInfo to null the GC thread can still be trying to call ForceGC
2. Occasionally the test will fail because it never sees any statics from the collectible module. I haven't been able to repro this, but the only way I can think of that it would fail is if the module is loaded and then unloaded while the GC thread is doing its 100 ms sleep. This changes the test so GCs are triggered for every module load and unload instead of occurring randomly.

### CODE REVIEW

**Comment by davmason** [comment]
Fixes #41652

File: src/tests/profiler/native/getappdomainstaticaddress/getappdomainstaticaddress.h [hunk 1]
```diff
@@ -52,18 +52,28 @@ class ManualEvent
                 spuriousCallback();
             }
         }
+        m_set = false;
     }
 
-    void Signal()
+    void WaitFor(int milliseconds, std::function<void()> spuriousCallback = DoNothing)
```
[the remark this answers is not in the document]
Comment (davmason) [reply src/tests/profiler/native/getappdomainstaticaddress/getappdomainstaticaddress.h]: The AutoEvent class is copy pasted from the dotnet-diagnostictests repo. I am not using the WaitFor method, but future tests might (tests in the old repo do and might be ported) and it's general enough to leave it in.

I moved the AutoEvent and ManualEvent to a separate header so it's clear they are general purpose classes and not specifically for this test.

**Comment by davmason** [comment]
@sywhang do you have any other issues with the test?

**Review by sywhang — APPROVED** [review]
LGTM

**Pull request merged**